fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -432,28 +433,64 @@ private ScoredPlan planFilter(@Nonnull PlanContext planContext, @Nonnull QueryCo
             boolean canSort = inExtractor.setSort(planContext.query.getSort(), planContext.query.isSortReverse());
-            if (!canSort && getConfiguration().shouldAttemptFailedInJoinAsOr()) {
-                // Can't implement as an in join because of the sort order. Try as an OR instead.
-                withInAsOr = planFilter(planContext, normalizeAndOrForInAsOr(inExtractor.asOr()));
+            if (!canSort) {
+                if (getConfiguration().shouldAttemptFailedInJoinAsUnion()) {
+                    withInAsOrUnion = planFilterWithInUnion(planContext, savedExtractor);
+                } else if (getConfiguration().shouldAttemptFailedInJoinAsOr()) {
+                    // Can't implement as an IN join because of the sort order. Try as an OR instead.
+                    QueryComponent asOr = normalizeAndOrForInAsOr(inExtractor.asOr());
+                    if (!filter.equals(asOr)) {
+                        withInAsOrUnion = planFilter(planContext, asOr);
+                    }
+                }
             }
         } else if (needOrdering) {
             inExtractor.sortByClauses();
         }
         final ScoredPlan withInJoin = planFilterWithInJoin(planContext, inExtractor, needOrdering);
-        if (withInAsOr != null) {
-            if (withInJoin == null || withInAsOr.score > withInJoin.score ||
-                    FieldWithComparisonCountProperty.evaluate(withInAsOr.plan) < FieldWithComparisonCountProperty.evaluate(withInJoin.plan)) {
-                return withInAsOr;
+        if (withInAsOrUnion != null) {
+            if (withInJoin == null || withInAsOrUnion.score > withInJoin.score ||
+                    FieldWithComparisonCountProperty.evaluate(withInAsOrUnion.plan) < FieldWithComparisonCountProperty.evaluate(withInJoin.plan)) {
+                return withInAsOrUnion;
             }
         }
         return withInJoin;
     }
 
     private ScoredPlan planFilterWithInJoin(@Nonnull PlanContext planContext, @Nonnull InExtractor inExtractor, boolean needOrdering) {
-        final QueryComponent filter = inExtractor.subFilter();
+        final ScoredPlan bestPlan = planFilterForInJoin(planContext, inExtractor.subFilter(), needOrdering);
+        if (bestPlan != null) {
+            final RecordQueryPlan wrapped = inExtractor.wrap(planContext.rankComparisons.wrap(bestPlan.plan, bestPlan.includedRankComparisons, metaData));
+            final ScoredPlan scoredPlan = new ScoredPlan(bestPlan.score, wrapped);
+            if (needOrdering) {
+                scoredPlan.planOrderingKey = inExtractor.adjustOrdering(bestPlan.planOrderingKey, false);
+            }
+            return scoredPlan;
+        }
+        return null;
+    }
+
+    private ScoredPlan planFilterWithInUnion(@Nonnull PlanContext planContext, @Nonnull InExtractor inExtractor) {
+        final ScoredPlan scoredPlan = planFilterForInJoin(planContext, inExtractor.subFilter(), true);
```

Comment:
what exactly does passing true here cause? Does the logic make sure that the query's sort requirement is applied (presumably looking at the passed in planContext?)

Reply:
`needOrdering` means that the method should populate the best plan's `planOrderingKey`.

The plan honors the query's sort requirement in both cases.